File: cli/src/agentobs/agents/cursor.py
"""Cursor has no native OTel export -- it needs the cursor-otel-hook shim
(vendored at integrations/cursor-otel-hook) wired into ~/.cursor/hooks.json.
Unlike Claude Code, this agent is configured via files, not shell env vars.
"""

from __future__ import annotations

import json
from pathlib import Path

HOOK_EVENTS = [
    "sessionStart",
    "sessionEnd",
    "preToolUse",
    "postToolUse",
    "postToolUseFailure",
    "beforeShellExecution",
    "afterShellExecution",
    "beforeMCPExecution",
    "afterMCPExecution",
    "beforeReadFile",
    "afterFileEdit",
    "beforeSubmitPrompt",
    "preCompact",
    "stop",
    "subagentStart",
    "subagentStop",
]


class CursorAgent:
    name = "cursor"

    def detect(self) -> bool:
        return (Path.home() / ".cursor").exists()

    def hooks_dir(self) -> Path:
        return Path.home() / ".cursor" / "hooks"

    def wrapper_script_path(self) -> Path:
        return self.hooks_dir() / "otel_hook.sh"

    def config_path(self) -> Path:
        return self.hooks_dir() / "otel_config.json"

    def hooks_json_path(self) -> Path:
        return Path.home() / ".cursor" / "hooks.json"

    def vendored_package_path(self) -> Path:
        # cli/src/agentobs/agents/cursor.py -> repo root is 4 parents up.
        return Path(__file__).resolve().parents[4] / "integrations" / "cursor-otel-hook"

    def otel_config(self, endpoint: str, mask_prompts: bool = False) -> dict:
        return {
            "OTEL_EXPORTER_OTLP_ENDPOINT": endpoint,
            "OTEL_SERVICE_NAME": "cursor-agent",
            "OTEL_EXPORTER_OTLP_PROTOCOL": "grpc",
            "OTEL_EXPORTER_OTLP_INSECURE": "true",
            "OTEL_EXPORTER_OTLP_HEADERS": None,
            "CURSOR_OTEL_MASK_PROMPTS": "true" if mask_prompts else "false",
            "OTEL_EXPORTER_OTLP_TIMEOUT": "30",
        }

    def hooks_json(self, existing: dict | None = None) -> dict:
        """Merge our hook entry into each event's list -- never replace whatever
        else (other tools, e.g. hindsight) already registered for that event."""
        wrapper = str(self.wrapper_script_path())
        entry = {"command": wrapper, "timeout": 5}

        merged = {"version": 1, "hooks": {}}
        if existing:
            merged["version"] = existing.get("version", 1)
            merged["hooks"] = {k: list(v) for k, v in existing.get("hooks", {}).items()}

        for event in HOOK_EVENTS:
            entries = merged["hooks"].setdefault(event, [])
            if not any(e.get("command") == wrapper for e in entries):
                entries.append(entry)

        return merged

    def wrapper_script(self) -> str:
        return (
            "#!/bin/bash\n"
            f'exec cursor-otel-hook --config "{self.config_path()}" "$@"\n'
        )

File: cli/internal/agents/cursor.go
package agents

import (
	"fmt"
	"os"
	"path/filepath"
)

var HookEvents = []string{
	"sessionStart",
	"sessionEnd",
	"preToolUse",
	"postToolUse",
	"postToolUseFailure",
	"beforeShellExecution",
	"afterShellExecution",
	"beforeMCPExecution",
	"afterMCPExecution",
	"beforeReadFile",
	"afterFileEdit",
	"beforeSubmitPrompt",
	"preCompact",
	"stop",
	"subagentStart",
	"subagentStop",
}

type CursorAgent struct{}

func (CursorAgent) Name() string { return "cursor" }

func (CursorAgent) Detect() bool {
	home, err := os.UserHomeDir()
	if err != nil {
		return false
	}
	_, err = os.Stat(filepath.Join(home, ".cursor"))
	return err == nil
}

func (CursorAgent) HooksDir() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, ".cursor", "hooks"), nil
}

func (a CursorAgent) WrapperScriptPath() (string, error) {
	dir, err := a.HooksDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(dir, "otel_hook.sh"), nil
}

func (a CursorAgent) ConfigPath() (string, error) {
	dir, err := a.HooksDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(dir, "otel_config.json"), nil
}

func (CursorAgent) HooksJSONPath() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, ".cursor", "hooks.json"), nil
}

func (CursorAgent) OtelConfig(endpoint string, maskPrompts bool, authToken string) map[string]interface{} {
	var headers interface{}
	if authToken != "" {
		headers = "Authorization=Bearer " + authToken
	}
	return map[string]interface{}{
		"OTEL_EXPORTER_OTLP_ENDPOINT": endpoint,
		"OTEL_SERVICE_NAME":           "cursor-agent",
		"OTEL_EXPORTER_OTLP_PROTOCOL": "grpc",
		"OTEL_EXPORTER_OTLP_INSECURE": "true",
		"OTEL_EXPORTER_OTLP_HEADERS":  headers,
		"CURSOR_OTEL_MASK_PROMPTS":    boolStr(maskPrompts),
		"OTEL_EXPORTER_OTLP_TIMEOUT":  "30",
	}
}

func boolStr(b bool) string {
	if b {
		return "true"
	}
	return "false"
}

// HooksJSON merges our hook entry into each event's list -- never replaces
// whatever else (other tools, e.g. a memory-retention plugin) already
// registered for that event, since Cursor supports multiple commands per event.
func (a CursorAgent) HooksJSON(existing map[string]interface{}) (map[string]interface{}, error) {
	wrapper, err := a.WrapperScriptPath()
	if err != nil {
		return nil, err
	}
	entry := map[string]interface{}{"command": wrapper, "timeout": 5}

	merged := map[string]interface{}{"version": 1}
	hooks := map[string]interface{}{}

	if existing != nil {
		if v, ok := existing["version"]; ok {
			merged["version"] = v
		}
		if existingHooks, ok := existing["hooks"].(map[string]interface{}); ok {
			for event, rawEntries := range existingHooks {
				if entries, ok := rawEntries.([]interface{}); ok {
					copied := make([]interface{}, len(entries))
					copy(copied, entries)
					hooks[event] = copied
				}
			}
		}
	}

	for _, event := range HookEvents {
		entries, _ := hooks[event].([]interface{})
		alreadyPresent := false
		for _, e := range entries {
			if m, ok := e.(map[string]interface{}); ok {
				if cmd, ok := m["command"].(string); ok && cmd == wrapper {
					alreadyPresent = true
					break
				}
			}
		}
		if !alreadyPresent {
			entries = append(entries, entry)
		}
		hooks[event] = entries
	}

	merged["hooks"] = hooks
	return merged, nil
}

func (a CursorAgent) WrapperScript() (string, error) {
	configPath, err := a.ConfigPath()
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("#!/bin/bash\nexec agentobs cursor-hook --config \"%s\" \"$@\"\n", configPath), nil
}
